cdap-watchdog-api/src/main/java/co/cask/cdap/api/metrics/MetricDeleteQuery.java

```diff
@@ -18,33 +18,42 @@
 
 import com.google.common.base.Joiner;
 import com.google.common.base.Objects;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Maps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.function.Predicate;
 
 /**
  * Query that specifies parameters to delete entries from {@link MetricStore}.
  */
 public class MetricDeleteQuery {
-
   private final long startTs;
   private final long endTs;
   private final Collection<String> metricNames;
   private final Map<String, String> sliceByTagValues;
+  private final Predicate<List<String>> tagPredicate;
 
   public MetricDeleteQuery(long startTs, long endTs, Collection<String> metricNames,
-                           Map<String, String> sliceByTagValues) {
+                           Map<String, String> sliceByTagValues, Predicate<List<String>> tagPredicate) {
     this.startTs = startTs;
     this.endTs = endTs;
     this.metricNames = metricNames;
-    this.sliceByTagValues = Maps.newHashMap(sliceByTagValues);
+    this.sliceByTagValues = new LinkedHashMap<>(sliceByTagValues);
+    this.tagPredicate = tagPredicate;
   }
 
-  public MetricDeleteQuery(long startTs, long endTs,
+  public MetricDeleteQuery(long startTs, long endTs, Collection<String> metricNames,
                            Map<String, String> sliceByTagValues) {
-    this(startTs, endTs, ImmutableList.<String>of(), sliceByTagValues);
+    this(startTs, endTs, metricNames, sliceByTagValues,
+         aggregates -> Collections.indexOfSubList(aggregates, new ArrayList<>(sliceByTagValues.keySet())) == 0);
```

Comment:
Why need to keep this constructor? Can you always use the one with Predicate?

Reply:
I keep it because we have to use this predicate logic everywhere we call this method. I feel like it is good to keep these in one place.